File: app/src/main/assets/videos/shared/XVideos.go
package shared

import (
	"net/http"
	"net/url"
	"strings"
)

func XVideos(uri string, proxy *url.URL) ([]byte, error) {
	b, err := getXVideosPage(uri, proxy)
	if err != nil {
		println(err)
		return nil, err
	}
	obj := make(map[string]interface{})

	title := SubstringBytes(b, []byte("setVideoTitle('"), []byte("');"))
	obj["title"] = string(title)

	var result []byte
	result = []byte(title)
	return result, nil
}

func getXVideosPage(uri string, proxy *url.URL) ([]byte, error) {
	result, err := Fetch(uri, nil, proxy, func(r *http.Request) {
		r.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9")
		r.Header.Set("Accept-Encoding", "gzip, deflate, br")
		r.Header.Set("Accept-Language", "en")
		r.Header.Set("Cache-Control", "no-cache")
		r.Header.Set("Connection", "keep-alive")
		r.Header.Set("Pragma", "no-cache")
		r.Header.Set("sec-ch-ua", "\"Google Chrome\";v=\"95\", \"Chromium\";v=\"95\", \";Not A Brand\";v=\"99\"")
		r.Header.Set("sec-ch-ua-mobile", "?0")
		r.Header.Set("sec-ch-ua-platform", "\"Windows\"")
		r.Header.Set("Sec-Fetch-Dest", "document")
		r.Header.Set("Sec-Fetch-Mode", "navigate")
		r.Header.Set("Sec-Fetch-Site", "none")
		r.Header.Set("Sec-Fetch-User", "?1")
		r.Header.Set("Upgrade-Insecure-Requests", "1")
		r.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.69 Safari/537.36")
	})
	if err != nil {
		return nil, err
	}
	return result, nil
}

// https://www.xvideos.com/video73139749/pretty_girl_having_great_fuck_from_her_boyfriend
func IsXVideosUri(u string) bool {
	return strings.HasPrefix("https://www.xvideos.com", u)
}

File: app/src/main/java/euphoria/psycho/fileserver/XVideos.java
package euphoria.psycho.fileserver;

import java.net.HttpURLConnection;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.Proxy.Type;
import java.net.URL;

import euphoria.psycho.fileserver.VideoDatabase.Video;

public class XVideos {
    public static Video fetch(String uri) throws Exception {
        HttpURLConnection c = (HttpURLConnection) new URL(uri).openConnection(new Proxy(Type.HTTP, new InetSocketAddress("127.0.0.1", 10809)));
        c.addRequestProperty("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9");
        c.addRequestProperty("Accept-Encoding", "gzip, deflate, br");
        c.addRequestProperty("Accept-Language", "en");
        c.addRequestProperty("Cache-Control", "no-cache");
        c.addRequestProperty("Connection", "keep-alive");
        c.addRequestProperty("Pragma", "no-cache");
        c.addRequestProperty("sec-ch-ua", "\"Google Chrome\";v=\"95\", \"Chromium\";v=\"95\", \";Not A Brand\";v=\"99\"");
        c.addRequestProperty("sec-ch-ua-mobile", "?0");
        c.addRequestProperty("sec-ch-ua-platform", "\"Windows\"");
        c.addRequestProperty("Sec-Fetch-Dest", "document");
        c.addRequestProperty("Sec-Fetch-Mode", "navigate");
        c.addRequestProperty("Sec-Fetch-Site", "none");
        c.addRequestProperty("Sec-Fetch-User", "?1");
        c.addRequestProperty("Upgrade-Insecure-Requests", "1");
        c.addRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.69 Safari/537.36");
        String contents = Shared.readString(c);
        Video video = new Video();
        video.Title = Shared.substring(contents, "setVideoTitle('", "');");
        video.Url = uri;
        video.Play = Shared.substring(contents, "setVideoUrlHigh('", "');");
        video.MusicPlay = "";
        video.MusicTitle = "";
        video.MusicAuthor = "";
        video.Cover = Shared.substring(contents, "setThumbUrl169('", "');");
        video.CreateAt = System.currentTimeMillis();
        video.UpdateAt = System.currentTimeMillis();
        return video;
    }
}
